contracts/modules/token/TokenController.sol

```diff
@@ -310,13 +311,68 @@ contract TokenController is ITokenController, LockHandler, MasterAwareV2 {
     }
   }
 
-  /**
-  * @dev Returns tokens locked for a specified address for a
-  *    specified reason
-  *
-  * @param _of The address whose tokens are locked
-  * @param _reason The reason to query the lock tokens for
-  */
+  /// @notice Withdraws NXM from the Nexus platform based on specified options.
+  /// @dev    Ensure the NXM is available and not locked before withdrawal. Only set flags in `WithdrawNxmOptions` for
+  ///         withdrawable NXM. Reverts if some of the NXM being withdrawn is locked or unavailable.
+  /// @param stakingPoolDeposit  Details for withdrawing from staking pools.
+  /// @param coverNotes          Details for withdrawing V1 cover notes.
+  /// @param batchSize           The maximum number of iterations to avoid unbounded loops when withdrawing
+  ///                            governance and/or assessment rewards.
+  /// @param withdrawOptions     Options specifying which NXM types to withdraw, set flags to true to include specific
+  ///                            withdrawable NXM.
+  function withdrawNXM(
+    StakingPoolDeposit calldata stakingPoolDeposit,
+    V1CoverNotes calldata coverNotes,
+    uint batchSize,
+    WithdrawNxmOptions calldata withdrawOptions
+  ) external whenNotPaused {
+
+    // assessment stake
+    if (withdrawOptions.assessmentStake) {
+      IAssessment _assessment = assessment();
+      (uint96 amount, , ) = _assessment.stakeOf(msg.sender);
+      _assessment.unstakeFor(msg.sender, amount, msg.sender);
+    }
+
+    // assessment and governance rewards
+    if (withdrawOptions.governanceRewards || withdrawOptions.assessmentRewards) {
+      withdrawPendingRewards(
+        msg.sender,
+        withdrawOptions.governanceRewards,
+        withdrawOptions.assessmentRewards,
+        batchSize
+      );
+    }
+
+    // staking pools rewards and stake (including manager rewards)
+    if (withdrawOptions.stakingPoolStake || withdrawOptions.stakingPoolRewards) {
+      for (uint i = 0; i < stakingPoolDeposit.tokenIds.length; i++) {
+        uint tokenId = stakingPoolDeposit.tokenIds[i];
+        uint[] memory trancheIds = stakingPoolDeposit.tokenTrancheIds[i];
+        uint poolId = stakingNFT.stakingPoolOf(tokenId);
+        stakingPool(poolId).withdraw(tokenId, withdrawOptions.stakingPoolStake, withdrawOptions.stakingPoolRewards, trancheIds);
+      }
+    }
+
+    // v1 cover notes
+    if (withdrawOptions.v1CoverNotes) {
+      withdrawCoverNote(msg.sender, coverNotes.coverIds, coverNotes.reasonIndexes);
+    }
+
+
+    // v1 pooled staking stake
+    if (withdrawOptions.v1PooledStakingStake) {
+      pooledStaking().withdrawForUser(msg.sender);
+    }
+
+    // v1 - always withdraw (no revert)
+    _withdrawClaimAssessmentTokensForUser(msg.sender);
+    pooledStaking().withdrawReward(msg.sender);
```

Comment:
Let's evict v1 stuff from this function so we don't have to change it at a later date (contracts+frontend).